presto-main/src/main/java/com/facebook/presto/operator/LookupJoinOperator.java

```diff
@@ -80,19 +108,50 @@ public OperatorContext getOperatorContext()
     @Override
     public List<Type> getTypes()
     {
-        return types;
+        return allTypes;
     }
 
     @Override
     public void finish()
     {
+        if (!finishing && lookupSourceSupplier.hasSpilled()) {
+            // Last LookupJoinOperator will handle spilled pages
+            int count = lookupJoinsCount.decrementAndGet();
+            checkState(count >= 0);
+            if (count > 0) {
+                spiller = Optional.empty();
+            }
+            else {
+                // last LookupJoinOperator might not spilled at all, but other parallel joins could
+                // so we have to load spiller if this operator hasn't used one yet
+                ensureSpillerLoaded();
+                unspillNextLookupSource();
```

Comment:
please review rewritten commit (and refactor commit which is at begin of the PR)